thoth-api/src/graphql/model.rs

```diff
@@ -1984,21 +1989,74 @@ impl MutationRoot {
+            } else {
+                location
+                    .update(&context.db, &data, &account_id)
+                    .map_err(|e| e.into())
+            }
+        } else {
+            // if changes to a non-canonical location don't result in it becoming canonical, perform a regular update.
+            if !data.canonical {
+                location
+                    .update(&context.db, &data, &account_id)
+                    .map_err(|e| e.into())
+            // if user changes a non-canonical location to canonical, perform two simultaneous updates:
+            // change the old canonical location to non-canonical, and change the old non-canonical location to canonical
+            } else {
+                let mut old_canonical_location_id: Option<Uuid> = None;
+                let connection = &mut context.db.get().unwrap();
+
+                let canonical_location = location::table
+                    .filter(location::publication_id.eq(data.publication_id))
+                    .filter(location::canonical.eq(true))
+                    .first::<Location>(&mut context.db.get()?);
+
+                let final_canonical_location = match canonical_location {
+                    Ok(location) => location,
+                    Err(e) => {
+                        return Err(ThothError::from(e).into());
+                    }
+                };
+
+                let old_canonical_location = Some(PatchLocation {
+                    location_id: final_canonical_location.location_id,
+                    publication_id: final_canonical_location.publication_id,
+                    landing_page: final_canonical_location.landing_page.clone(),
+                    full_text_url: final_canonical_location.full_text_url.clone(),
+                    location_platform: final_canonical_location.location_platform.clone(),
+                    canonical: false,
+                });
+                if let Some(ref old_canonical_location) = old_canonical_location {
+                    old_canonical_location_id = Some(old_canonical_location.location_id);
+                }
+
+                if let Some(old_canonical_location_id) = old_canonical_location_id {
+                    let _ = connection.transaction(|connection| {
+                        // Update the current canonical location to non-canonical
+                        diesel::update(location::table.find(old_canonical_location_id))
+                            .set(old_canonical_location)
+                            .execute(connection)?;
+
+                        // Update the current non-canonical location to canonical
+                        diesel::update(location::table.find(data.location_id))
+                            .set(&data)
+                            .execute(connection)?;
+
+                        Ok::<_, diesel::result::Error>(())
+                    });
+                }
+                Ok(location)
```

Comment:
I've refactored this; see `update_canonical_location()` in `location/crud.rs`. Now I have it just returning `Ok(())`, since I'm not sure what I should be returning from this function?

Reply:
Refactored this significantly, so I'm going to resolve this comment and you can take a look at the new version.